### PR TITLE
[EAGLE] perf: Fuse draft topk1, vocab embedding, and capture staging

### PR DESCRIPTION
**Pull request by mattteochen**
## Motivation

This MTP draft optimization branch targets three repeated costs inside EAGLE/MTP speculative decoding under CUDA graph replay:

1. **`topk=1` draft postprocess**: PyTorch eager `argmax` over 129k-151k draft logits has too little parallelism at small decode batch sizes, and the chain-shaped draft tree still paid extra tensor/list materialization.
2. **Draft-model vocab-parallel embedding**: the TP local embedding path emitted separate mask pointwise kernels, `F.embedding`/`indexSelectSmallIndex`, and `masked_fill_kernel` for every draft forward.
3. **Draft capture eager input staging**: multi-step draft CUDA graph capture routes per-step forwards through the eager runner; staging graph-stable draft buffers into the eager registry records redundant copy nodes in the captured graph.

The goal is to reduce draft-loop launch count and remove redundant copies without changing the speculative decoding semantics.

### 1. Fuse `topk=1` draft postprocess

- Add `speculative/triton_ops/topk1.py` with a split-reduction argmax path (`draft_topk1_postprocess`).
- Split the vocab across CTAs, then finalize the winning split with torch-compatible argmax tie-breaking.
- Fuse position advance and, on the CUDA fast path, the draft-token column store into the finalize kernel.
- In `eagle_worker_v2.draft_forward`, special-case `topk == 1` as a chain:
  - materialize the `(bs, num_steps)` draft-token matrix directly,
  - skip `select_top_k_tokens`, per-step token lists, and post-loop `torch.cat`,
  - keep non-CUDA / rejection-sampling fallback behavior intact.

### 2. Fuse TP vocab-parallel embedding in the draft model

- Add a fused Triton vocab-parallel embedding kernel for the standard unquantized path.
- The fused kernel directly writes zero rows for out-of-shard token ids and gathers local rows for in-shard ids.
- Enable the fused path only when the safe conditions hold:
  - `tp_size > 1`,
  - CUDA contiguous int32/int64 token ids,
  - unquantized embedding method,
  - CUDA fp16/bf16/fp32 2D weight with contiguous columns.
- Keep existing PyTorch/compiled fallback for CPU/NPU, TP1, quantized/custom embedding methods, unsupported layouts, or when disabled.
- Add kill-switch: `SGLANG_OPT_USE_TRITON_VOCAB_PARALLEL_EMBEDDING=0`.

### 3. Remove redundant eager input staging during draft CUDA graph capture

- Add `ForwardBatch.skip_eager_input_staging` to explicitly mark batches backed by persistent graph buffers.
- Set this flag for EAGLE draft and frozen-KV MTP draft CUDA graph capture batches.
- Change `EagerRunner.load_batch` so generic capture mode still stages by default; no-copy is used only for:
  - the global `SGLANG_EAGER_INPUT_NO_COPY` override, or
  - an explicit `skip_eager_input_staging` batch.
- Rename the capture-specific env to `SGLANG_EAGER_INPUT_NO_COPY_IN_DRAFT_CAPTURE`.

Why this is safe for draft capture: those `ForwardBatch` fields are views into the draft graph runner's persistent `self.buffers`; replay refreshes contents in-place, so captured tensor addresses remain stable. Generic capture mode alone does not prove that aliasing contract, so it continues to stage.

## Accuracy Tests

Server:
```bash
sglang serve \
  --model-path zai-org/GLM-5.2-FP8 \
  --tp 8 \
  --mem-fraction-static 0.85 \
  --max-running-requests 256 \
  --host 0.0.0.0
```
main:
```bash
== aime25 ==
30 examples x 16 repeats  |  1804.6s  |  5083 tok/s  |  9.2M tokens

* pass@1[avg-of-16]  =  90.00% +/- 4.04% (SEM 1.01%)
  pass@16            =  96.67%
  majority@16        =  95.00%
  no_answer          =  0.42%
  stop_rate          =  100.00%
  truncated_rate     =  0.00%
  error_rate         =  0.00%
```

PR:
```bash
== aime25 ==
30 examples x 16 repeats  |  1899.1s  |  4891 tok/s  |  9.3M tokens

* pass@1[avg-of-16]  =  90.62% +/- 4.90% (SEM 1.23%)
  pass@16            =  100.00%
  majority@16        =  93.33%
  no_answer          =  1.04%
  stop_rate          =  99.79%
  truncated_rate     =  0.21%
  error_rate         =  0.00%
```

At steady state (max conc 64 decode)

Metric | main | pr | Δ (pr − main)
-- | -- | -- | --
Steady-state samples | 735 | 726 | —
Mean accept len | 4.133 | 4.112 | −0.021 (−0.5%)

### GLM5.2-FP8 B300
MTP config:
```bash
  --speculative-algorithm EAGLE \
  --speculative-num-steps 5 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 6
```

| Section | main | PR | speedup |
| --- | ---: | ---: | ---: |
| `draft` GPU range total | 1.138ms | 0.958ms | 1.19x |
| `draft_postprocess` + `vocab` GPU range total | 60.7us | 13.9us | 4.37x |

main:
Screenshot 2026-07-08 at 12 12 32

PR:
Screenshot 2026-07-08 at 12 12 48

B300 1k1k pareto:
image

```bash
series                            conc       TPS/user (sp)    out TPS/GPU (sp)  total TPS/GPU (sp)        TTFT ms (sp)
----------------------------------------------------------------------------------------------------------------------
base                                 4     282.18 ( 1.00x)     118.91 ( 1.00x)     239.36 ( 1.00x)     255.70 ( 1.00x)
base                                 8     195.21 ( 1.00x)     171.52 ( 1.00x)     344.08 ( 1.00x)     263.47 ( 1.00x)
base                                16     132.82 ( 1.00x)     240.06 ( 1.00x)     476.58 ( 1.00x)     274.72 ( 1.00x)
base                                32      85.75 ( 1.00x)     318.68 ( 1.00x)     639.07 ( 1.00x)     286.08 ( 1.00x)
base                                64      56.68 ( 1.00x)     415.07 ( 1.00x)     831.27 ( 1.00x)     304.26 ( 1.00x)
fused                                4     284.06 ( 1.01x)     121.05 ( 1.02x)     243.68 ( 1.02x)     252.18 ( 1.01x)
fused                                8     198.17 ( 1.02x)     177.69 ( 1.04x)     356.45 ( 1.04x)     260.40 ( 1.01x)
fused                               16     134.48 ( 1.01x)     242.92 ( 1.01x)     482.27 ( 1.01x)     272.85 ( 1.01x)
fused                               32      88.22 ( 1.03x)     328.65 ( 1.03x)     659.07 ( 1.03x)     287.10 ( 1.00x)
fused                               64      56.92 ( 1.00x)     432.44 ( 1.04x)     866.05 ( 1.04x)     300.44 ( 1.01x)
```

### DSV4-Flash B300
MTP config:
```bash
  --speculative-algorithm EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4
```
main:
image

PR:
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). N/A: internal kernel fast path with env kill-switch.
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29209821826](https://github.com/sgl-project/sglang/actions/runs/29209821826)
Latest PR Test (Extra): :x: [Run #29209821764](https://github.com/sgl-project/sglang/actions/runs/29209821764)

### CODE REVIEW

**Review by kpham-sgl — COMMENTED** [review]
> - topk=1 draft postprocess: PyTorch eager argmax over 129k-151k draft logits has too little parallelism at small decode batch sizes, and the chain-shaped draft tree still paid extra tensor/list materialization.
> - Draft-model vocab-parallel embedding: the TP local embedding path emitted separate mask pointwise kernels, F.embedding/indexSelectSmallIndex, and masked_fill_kernel for every draft forward.
> - Draft capture eager input staging: multi-step draft CUDA graph capture routes per-step forwards through the eager runner; staging graph-stable draft buffers into the eager registry records redundant copy nodes in the captured graph.

@mattteochen can you split this PR into 3 parts

**Comment by mattteochen** [comment]
Hi @kpham-sgl , I've split in https://github.com/sgl-project/sglang/pull/30947, https://github.com/sgl-project/sglang/pull/30948, https://github.com/sgl-project/sglang/pull/30949
